images/encryption/encryption.go

```diff
@@ -0,0 +1,332 @@
+// of the given keywrapper with it and returns the result in case the Unwrap functions
+// does not return an error. If all attempts fail, an error is returned.
+func preUnwrapKey(keywrapper keywrap.KeyWrapper, dc *config.DecryptConfig, b64Annotations string) ([]byte, error) {
+	if b64Annotations == "" {
+		return nil, nil
+	}
+	for _, b64Annotation := range strings.Split(b64Annotations, ",") {
+		annotation, err := base64.StdEncoding.DecodeString(b64Annotation)
+		if err != nil {
+			return nil, errors.New("Could not base64 decode the annotation")
+		}
+		optsData, err := keywrapper.UnwrapKey(dc, annotation)
+		if err != nil {
+			continue
+		}
+		return optsData, nil
+	}
+	return nil, errors.New("No suitable key found for decrypting layer key")
+}
+
+// commonEncryptLayer is a function to encrypt the plain layer using a new random
+// symmetric key and return the LayerBlockCipherHandler's JSON in string form for
+// later use during decryption
+func commonEncryptLayer(plainLayerReader io.ReaderAt, symKey []byte, typ blockcipher.LayerCipherType) (content.ReaderDigester, []byte, error) {
+	opts := blockcipher.LayerBlockCipherOptions{
+		SymmetricKey: symKey,
+	}
+	lbch, err := blockcipher.NewLayerBlockCipherHandler()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	encLayerReader, opts, err := lbch.Encrypt(plainLayerReader, typ, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	optsData, err := json.Marshal(opts)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "Could not JSON marshal opts")
+	}
+
+	return encLayerReader, optsData, err
+}
+
+// commonDecryptLayer decrypts an encrypted layer previously encrypted with commonEncryptLayer
+// by passing along the optsData
+func commonDecryptLayer(encLayerReader io.ReaderAt, optsData []byte) (content.ReaderDigester, error) {
+	opts := blockcipher.LayerBlockCipherOptions{}
+	err := json.Unmarshal(optsData, &opts)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Could not JSON unmarshal optsData")
+	}
+
+	lbch, err := blockcipher.NewLayerBlockCipherHandler()
+	if err != nil {
+		return nil, err
+	}
+
+	plainLayerReader, opts, err := lbch.Decrypt(encLayerReader, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return plainLayerReader, nil
+}
+
+// GetCryptoConfigFromAnnotations expects the dcparameters in json format in
+// the descriptor's Annotations
+func GetCryptoConfigFromAnnotations(desc *ocispec.Descriptor) (*config.CryptoConfig, error) {
+	var err error
+
+	cc := &config.CryptoConfig{
+		Dc: &config.DecryptConfig{},
+	}
+
+	dcparametersStr := desc.Annotations["_dcparameters"]
+	if len(dcparametersStr) == 0 {
+		return cc, nil
+	}
+	delete(desc.Annotations, "_dcparameters")
+
+	cc.Dc.Parameters, err = DCParametersFromJSON(dcparametersStr)
+	return cc, err
+}
+
+// DCParametersFromJSON converts a json representation of the dcparameters map into a native map
+func DCParametersFromJSON(dcparametersStr string) (map[string][][]byte, error) {
```

Comment:
Golang JSON already marshalls and unmarshals bytes to base64 encoded strings, so this should not be necessary at all.

Reply:
I kept the two functions but simplified them: https://github.com/containerd/containerd/pull/3001/commits/c05cb186fa115cfc294463fc84ba0cdbe1752931